internal/getproviders/package_location_http_archive.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package getproviders
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/hashicorp/go-getter"
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/opentofu/opentofu/internal/httpclient"
+	"github.com/opentofu/opentofu/internal/logging"
+)
+
+// PackageHTTPURL is a provider package location accessible via HTTP.
+//
+// Its value is a URL string using either the http: scheme or the https: scheme.
+// The URL should respond with a .zip archive whose contents are to be extracted
+// into a local package directory.
+type PackageHTTPURL string
+
+var _ PackageLocation = PackageHTTPURL("")
+
+func (p PackageHTTPURL) String() string { return string(p) }
+
+func (p PackageHTTPURL) InstallProviderPackage(ctx context.Context, meta PackageMeta, targetDir string, allowedHashes []Hash) (*PackageAuthenticationResult, error) {
+	url := meta.Location.String()
+
+	// When we're installing from an HTTP URL we expect the URL to refer to
+	// a zip file. We'll fetch that into a temporary file here and then
+	// delegate to installFromLocalArchive below to actually extract it.
+	// (We're not using go-getter here because its HTTP getter has a bunch
+	// of extraneous functionality we don't need or want, like indirection
+	// through X-Terraform-Get header, attempting partial fetches for
+	// files that already exist, etc.)
+
+	retryableClient := retryablehttp.NewClient()
+	retryableClient.HTTPClient = httpclient.New()
+	retryableClient.RetryMax = maxHTTPPackageRetryCount
+	retryableClient.RequestLogHook = func(logger retryablehttp.Logger, _ *http.Request, i int) {
+		if i > 0 {
+			logger.Printf("[INFO] failed to fetch provider package; retrying")
+		}
+	}
+	retryableClient.Logger = log.New(logging.LogOutput(), "", log.Flags())
+
+	req, err := retryablehttp.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("invalid provider download request: %w", err)
+	}
+	resp, err := retryableClient.Do(req)
+	if err != nil {
+		if ctx.Err() == context.Canceled {
+			// "context canceled" is not a user-friendly error message,
+			// so we'll return a more appropriate one here.
+			return nil, fmt.Errorf("provider download was interrupted")
+		}
+		return nil, fmt.Errorf("%s: %w", HostFromRequest(req.Request), err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unsuccessful request to %s: %s", url, resp.Status)
+	}
+
+	f, err := os.CreateTemp("", "terraform-provider")
```

Comment:
Would you mind adding a comment here to explain why we're using the name `terraform-provider`. 

I have concerns that someone one day will come along and do a large find+replace to OpenTofu and not know wether to change this or not. 

In this case I think it should stay as `terraform-provider` because it is terraform providers that we're downloading.

Reply:
As far as I know this naming convention doesn't actually matter at all, because this file only exists on disk briefly so that we can verify its checksum and then extract it into the target directory without having to read the entire (potentially-quite-large) `.zip` file into a memory buffer.

The filepath generated here is therefore observed only by the `PackageLocalArchive.InstallProviderPackage` call this delegates to below, and then the file is deleted before we return.

I'm not going to change this now for the reasons in https://github.com/opentofu/opentofu/pull/2579#issuecomment-2718378264, but I believe we _could_ change this to be any prefix we like in a future PR if we want to remove the reference to our predecessor project.